### PR TITLE
TST: groupby apply called multiple times

### LINKED ISSUE #31111: Strange groupby apply behaviour
#### Minimal reproducible example

```python
import pandas as pd
import numpy as np

tdf = pd.DataFrame({"tree": [0, 0, 0, 0, 1, 1, 1, 1],
                    "into": ["0-2", "0-4", "0-10", np.nan, "1-2", "1-3", "1-7", np.nan],
                    "leaf_value": [0,0,0,3,0,0,0,4]},
                   index=["0-0", "0-2", "0-4", "0-10", "1-0", "1-1", "1-2", "1-7"])

def deduce_tree(df):
    print("DEDUCING TREE WITH INDEX:\n",df.index)
    next_id = df.index[0]
    while isinstance(next_id, str):
        print(next_id)
        next_node = df.loc[next_id, :]
        next_id = df.loc[next_id, "into"]
    print("RETURNING:\n", next_node)
    return next_node

tdf.groupby("tree").apply(deduce_tree)

```
<details>

DEDUCING GREE WITH INDEX:
 Index(['0-0', '0-2', '0-4', '0-10'], dtype='object')
0-0
0-2
0-4
0-10
RETURNING:
 tree            0
into          NaN
leaf_value      3
Name: 0-10, dtype: object
DEDUCING GREE WITH INDEX:
 Index(['1-0', '1-1', '1-2', '1-7'], dtype='object')
1-0
DEDUCING GREE WITH INDEX:
 Index(['0-0', '0-2', '0-4', '0-10'], dtype='object')
0-0
0-2
0-4
0-10
RETURNING:
 tree            0
into          NaN
leaf_value      3
Name: 0-10, dtype: object
DEDUCING GREE WITH INDEX:
 Index(['1-0', '1-1', '1-2', '1-7'], dtype='object')
1-0
1-2
1-7
RETURNING:
 tree            1
into          NaN
leaf_value      4
Name: 1-7, dtype: object

</details>

#### Problem description

This came up when trying to analyze a boosted tree internals. As the apply function gets called with a print statement, when it gets to the line `next_node = df.loc[next_id, :]` it just calls the deduce_tree function again, with the group 0.
it prints out `DEDUCING TREE WITH INDEX` 3 times as opposed to 2, and for some reason interrupts the function.

the result of `tdf.groupby("tree").apply(deduce_tree)` is correct, but it seems to do some unnecessary work and if I want to implement some side effects into `deduce_tree` it gets messed up.

Can anyone explain why it works like this? Is this some bug? How can a .loc interrupt a function?

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.3.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.2.1-1.el7.elrepo.x86_64
machine          : x86_64
processor        : 
byteorder        : little
LC_ALL           : None
LANG             : C.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 0.25.3
numpy            : 1.18.1
pytz             : 2018.4
dateutil         : 2.7.3
pip              : 19.2.3
setuptools       : 41.0.1
Cython           : None
pytest           : 4.1.1
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : 1.0.1
pymysql          : None
psycopg2         : 2.8.1 (dt dec pq3 ext lo64)
jinja2           : 2.10
IPython          : 7.8.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : 3.1.2
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : 1.2.8
tables           : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None

</details>

### PR DESCRIPTION
**Pull request by Rohith295**
closes #31111 
xref https://github.com/pandas-dev/pandas/pull/24748
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ Added test for testing func should be called len(set(groupby_column)) times when passed to groupby.apply(func) ] whatsnew entry

### CODE REVIEW

File: pandas/tests/groupby/test_apply.py [hunk 1]
```diff
@@ -974,3 +974,21 @@ def test_apply_function_with_indexing_return_column():
     result = df.groupby("foo1", as_index=False).apply(lambda x: x.mean())
     expected = DataFrame({"foo1": ["one", "three", "two"], "foo2": [3.0, 4.0, 4.0]})
     tm.assert_frame_equal(result, expected)
+
+
+def test_apply_function_called_count(capsys):
+    # GH: 31111
+    # groupby-apply need to execute len(set(group_by_columns)) times
+    # `https://github.com/pandas-dev/pandas/issues/31111`
```
Comment (MarcoGorelli) [inline pandas/tests/groupby/test_apply.py:982]: If you already have the issue number you don't need the link as well

File: pandas/tests/groupby/test_apply.py [hunk 2]
```diff
@@ -974,3 +974,21 @@ def test_apply_function_with_indexing_return_column():
...
+    
+
+    function_called_count = 2 # Number of times `apply` should call a function for the current test
```
Comment (MarcoGorelli) [inline pandas/tests/groupby/test_apply.py:985]: Perhaps call this `expected`, and call `capsys.readouterr().out.count("function_called")` `result`, and at the end do `assert result == expected`?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Thanks @Rohith295 ! Seems there's some linting failures, can you run `black` and `flake8` against your changed file?

File: pandas/tests/groupby/test_apply.py (hunk 1 above)
Comment (Rohith295) [reply pandas/tests/groupby/test_apply.py:982]: Alright. I understood that. Will make sure not to repeat this again.

File: pandas/tests/groupby/test_apply.py (hunk 2 above)
Comment (Rohith295) [reply pandas/tests/groupby/test_apply.py:985]: Good point. I will make it more readable.

File: pandas/tests/groupby/test_apply.py (hunk 1 above)
Comment (Rohith295) [reply pandas/tests/groupby/test_apply.py]: Changed accordingly. @MarcoGorelli

File: pandas/tests/groupby/test_apply.py [hunk 3]
```diff
@@ -974,3 +974,24 @@ def test_apply_function_with_indexing_return_column():
     result = df.groupby("foo1", as_index=False).apply(lambda x: x.mean())
     expected = DataFrame({"foo1": ["one", "three", "two"], "foo2": [3.0, 4.0, 4.0]})
     tm.assert_frame_equal(result, expected)
+
+
+def test_apply_function_called_count(capsys):
```
Comment (jreback) [inline pandas/tests/groupby/test_apply.py:979]: move this near test_group_apply_once_per_group and indicate the same issue number (as well)

Comment (jreback) [reply pandas/tests/groupby/test_apply.py:979]: and rename these test to test_group_apply_once_per_group2

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
this is very similar to these issues
# GH2936, GH7739, GH10519, GH2656, GH12155, GH20084, GH21417

File: pandas/tests/groupby/test_apply.py (hunk 3 above)
Comment (Rohith295) [reply pandas/tests/groupby/test_apply.py:979]: Alright @jreback . Will change it accordingly

Comment (Rohith295) [reply pandas/tests/groupby/test_apply.py:979]: Should remove `GH: 31111` and use `# GH2936, GH7739, GH10519, GH2656, GH12155, GH20084, GH21417` @jreback

Comment (jreback) [reply pandas/tests/groupby/test_apply.py:979]: you can add this issue number as well is ok

Comment (Rohith295) [reply pandas/tests/groupby/test_apply.py]: Alright. fixed accordingly

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm ping on green.

**Comment by jreback** [comment]
thanks @Rohith295

**Pull request merged**